### PR TITLE
Update routes add pagination

### PR DESCRIPTION
**Pull request by Cempl**

### CODE REVIEW

File: sdk/consts.go [hunk 1]
```diff
@@ -128,12 +129,15 @@ const (
...
-	transactionsRoute                 = "/transaction"
-	transactionRoute                  = "/transaction/%s"
-	transactionStatusRoute            = "/transaction/%s/status"
-	transactionsStatusRoute           = "/transaction/statuses"
-	announceAggregateRoute            = "/transaction/partial"
-	announceAggregateCosignatureRoute = "/transaction/cosignature"
+	transactionsRoute                 = "/transactions"
+	transactionsStatusPluralRoute     = "/transactionStatus"
+	transactionStatusSingularRoute    = "/transactionStatus/%s"
+	confirmedTransactionsRoute        = "/transactions/confirmed"
```
Comment (xgreenx) [inline sdk/consts.go:135]: You don't need to have 3 different routes for each group of transactions.
You are processing all rotes by the same way, you only are using the different urls(Different part in this url).
We already have example how to solve it.
image

File: sdk/consts.go [hunk 2]
```diff
@@ -128,12 +129,15 @@ const (
...
+	confirmedTransactionRoute         = "/transactions/confirmed/%s"
+	unconfirmedTransactionRoute       = "/transactions/unconfirmed/%s"
```
Comment (xgreenx) [inline sdk/consts.go:137]: Also seems you lost 1 route here=D

File: sdk/transaction_model.go [hunk 3]
```diff
@@ -118,6 +118,78 @@ type abstractTransactionDTO struct {
 	Signer    string                  `json:"signer"`
 }
 
+type transactionsPageDTO struct {
```
Comment (xgreenx) [inline sdk/transaction_model.go:121]: All DTO must be in *_internal.go file

File: sdk/transaction.go [hunk 4]
```diff
@@ -101,19 +319,19 @@ func (txs *TransactionService) GetTransactionStatus(ctx context.Context, id stri
...
 
 	dtos := transactionStatusDTOs(make([]*transactionStatusDTO, len(hashes)))
-	resp, err := txs.client.doNewRequest(ctx, http.MethodPost, transactionsStatusRoute, txIds, &dtos)
+	resp, err := txs.client.doNewRequest(ctx, http.MethodPost, transactionsStatusPluralRoute, txIds, &dtos)
 	if err != nil {
 		return nil, err
 	}
 
-	if err = handleResponseStatusCode(resp, map[int]error{404: ErrResourceNotFound, 409: ErrArgumentNotValid}); err != nil {
+	if err = handleResponseStatusCode(resp, map[int]error{400: ErrInvalidRequest, 409: ErrArgumentNotValid}); err != nil {
```
Comment (xgreenx) [inline sdk/transaction.go:334]: Why did you remove process of 404 error?

File: sdk/transaction_test.go [hunk 5]
```diff
@@ -187,9 +255,195 @@ func TestTransactionService_GetTransaction_TransferTransaction(t *testing.T) {
...
+	assert.Nilf(t, err, "TransactionService.GetConfirmedTransactionsByIds returned error: %v", err)
+
+	tests.AssertEqual(t, 1, len(txs))
+
+	tests.ValidateStringers(t, confirmedTransactions.Transactions[0], txs[0])
+}
+
+func TestTransactionService_GetUnconfirmedTransaction(t *testing.T) {
+	mockServer.AddRouter(&mock.Router{
+		Path:     fmt.Sprintf("/transactions/unconfirmed/%s", transactionId),
```
Comment (xgreenx) [inline sdk/transaction_test.go:326]: Let's use constants of url instead of strings.

**Review by xgreenx — CHANGES_REQUESTED** [review]
Are you sure that you able to build all code and all tests?=)

File: sdk/sdk_model.go [hunk 6]
```diff
@@ -84,6 +84,30 @@ type AccountTransactionsOption struct {
...
+)
+
+func (sD SortDirection) String() string {
+	return [...]string{"asc", "desc"}[sD]
+}
+
+type PaginationOrderingOptions struct {
+	PageSize      uint64 `url:"pageSize,omitempty"`
+	PageNumber    uint64 `url:"pageNumber,omitempty"`
+	Offset        string `bson:"offset,omitempty"`
```
Comment (xgreenx) [inline sdk/sdk_model.go:106]: why bson?

File: sdk/storage_internal.go [hunk 7]
```diff
@@ -338,3 +338,40 @@ func (ref *downloadInfoDTOs) toStruct(networkType NetworkType) ([]*DownloadInfo,
...
+		TotalEntries uint64 `json:"totalEntries"`
+		PageNumber   uint64 `json:"pageNumber"`
+		PageSize     uint64 `json:"pageSize"`
+		TotalPages   uint64 `json:"totalPages"`
+	} `json:"pagination"`
+}
+
+func (t *drivesPageDTO) toStruct(networkType NetworkType) (*DrivesPage, error) {
+	page := &DrivesPage{
+		Drives: make([]Drive, len(t.Drives)),
```
Comment (xgreenx) [inline sdk/storage_internal.go:355]: Better to have []*Drive

File: sdk/transaction.go [hunk 8]
```diff
@@ -9,22 +9,18 @@ import (
...
-// returns Transaction for passed transaction id or hash
-func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (Transaction, error) {
+// returns Transaction for passed transaction group and hash
+func (txs *TransactionService) getTransaction(ctx context.Context, group string, id string) (Transaction, error) {
 	var b bytes.Buffer
 
-	url := net.NewUrl(fmt.Sprintf(transactionRoute, id))
-
-	resp, err := txs.client.doNewRequest(ctx, http.MethodGet, url.Encode(), nil, &b)
+	resp, err := txs.client.doNewRequest(ctx, http.MethodGet, fmt.Sprintf("%s/%s/%s", transactionsRoute, group, id), nil, &b)
```
Comment (xgreenx) [inline sdk/transaction.go:23]: Instead "%s/%s/%s" you can create template const string for that=D

File: sdk/transaction.go [hunk 9]
```diff
@@ -36,6 +32,101 @@ func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (T
 	return MapTransaction(&b, txs.client.GenerationHash())
 }
 
+// returns an array of Transaction's for passed array of transaction ids or hashes
+func (txs *TransactionService) getTransactionsByGroup(ctx context.Context, group TransactionGroups, tpOpts *TransactionsPageOptions) (*TransactionsPage, error) {
+	tspDTO := &transactionsPageDTO{}
+
+	u, err := addOptions(fmt.Sprintf("%s/%s", transactionsRoute, group.String()), tpOpts)
```
Comment (xgreenx) [inline sdk/transaction.go:39]: Instead of "%s/%s" you can have template=)

File: sdk/transaction.go [hunk 10]
```diff
@@ -72,7 +163,7 @@ func (txs *TransactionService) AnnounceAggregateBonded(ctx context.Context, tx *
 		tx.Payload,
 		tx.Hash.String(),
 	}
-	return txs.announceTransaction(ctx, &dto, announceAggregateRoute)
+	return txs.announceTransaction(ctx, &dto, fmt.Sprintf("%s/%s", transactionsRoute, partial.String()))
```
Comment (xgreenx) [inline sdk/transaction.go:166]: Let's you const string)

File: sdk/transaction.go [hunk 11]
```diff
@@ -36,6 +32,101 @@ func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (T
...
+	trS, err := txs.GetTransactionStatus(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	return txs.getTransaction(ctx, trS.Group, id)
+}
+
+// returns confirmed Transaction by id or hash
+func (txs *TransactionService) GetConfirmedTransaction(ctx context.Context, id string) (Transaction, error) {
```
Comment (xgreenx) [inline sdk/transaction.go:86]: I think we don't need this function and all functions below

File: sdk/transaction.go [hunk 12]
```diff
@@ -36,6 +32,101 @@ func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (T
...
+
+	if err = handleResponseStatusCode(resp, map[int]error{404: ErrResourceNotFound, 409: ErrArgumentNotValid}); err != nil {
+		return nil, err
+	}
+
+	return MapTransactions(&b, txs.client.GenerationHash())
+}
+
+// returns Transaction for passed transaction id or hash
+func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (Transaction, error) {
```
Comment (xgreenx) [inline sdk/transaction.go:76]: *GetAnyTransactionById will be better

File: sdk/transaction.go [hunk 13]
```diff
@@ -36,6 +32,101 @@ func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (T
...
+	return txs.getTransactionsByGroup(ctx, partial, tpOpts)
+}
+
+// returns an array of Transaction's for passed array of transaction ids or hashes
+func (txs *TransactionService) GetPartialTransactionsByIds(ctx context.Context, ids []string) ([]Transaction, error) {
+	return txs.getTransactionsByIds(ctx, partial, ids)
+}
+
 // returns an array of Transaction's for passed array of transaction ids or hashes
 func (txs *TransactionService) GetTransactions(ctx context.Context, ids []string) ([]Transaction, error) {
```
Comment (xgreenx) [inline sdk/transaction.go:131]: Maybe we don't need it, because you already added a new one?=)

File: sdk/transaction.go (hunk 8 above)
Comment (Cempl) [reply sdk/transaction.go:23]: Which template did you mean? This [one](https://golang.org/pkg/text/template/#pkg-overview) or?

File: sdk/storage_internal.go (hunk 7 above)
Comment (Cempl) [reply sdk/storage_internal.go:355]: I do not hink so, because we are missing pagination details in this way =))

File: sdk/sdk_model.go (hunk 6 above)
Comment (Cempl) [reply sdk/sdk_model.go:106]: If it will set as 'url' then exception generated on the rest-server side **""body":{"code":"InvalidArgument","message":"offset is not a valid object id"}"**. 
Offset on rest-server side looks like objectId for mongoDB(bson format inside). That is a reason why I used bson.

Comment (xgreenx) [reply sdk/sdk_model.go:106]: Okey
image

File: sdk/storage_internal.go (hunk 7 above)
Comment (xgreenx) [reply sdk/storage_internal.go:355]: Which details do you mean?

File: sdk/transaction.go (hunk 8 above)
Comment (xgreenx) [reply sdk/transaction.go:23]: Like this)

File: sdk/storage_internal.go (hunk 7 above)
Comment (Cempl) [reply sdk/storage_internal.go:355]: `TotalEntries uint64
	PageNumber   uint64
	PageSize     uint64
	TotalPages   uint64`

Comment (xgreenx) [reply sdk/storage_internal.go:355]: I meant []Drive -> []*Drive

File: sdk/transaction.go (hunk 13 above)
Comment (Cempl) [reply sdk/transaction.go:131]: By my point of view a lot of small functions better =))

File: sdk/transaction.go (hunk 12 above)
Comment (Wondertan) [reply sdk/transaction.go:76]: Note: That's for SDK semantics to be consistent
From Golang philosophy perspective, your original name is more preferable

File: sdk/transaction.go (hunk 13 above)
Comment (Cempl) [reply sdk/transaction.go:131]: From my point of view a lot of small functions better =))

Comment (Wondertan) [reply sdk/transaction.go:136]: @Cempl so where are filters and paginations ?

File: sdk/transaction.go [hunk 14]
```diff
@@ -36,6 +32,106 @@ func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (T
 	return MapTransaction(&b, txs.client.GenerationHash())
 }
 
+// returns an array of Transaction's for passed array of transaction ids or hashes
```
Comment (Wondertan) [inline sdk/transaction.go:35]: Golang's commenting style requires always starting comments with the name of a method. Pls, change that here and in other methods as well.

Comment (Wondertan) [reply sdk/transaction.go:35]: Note: this rule applies only to public methods/funcs, thus others who start with a small letter might have any, likely helpful, comment.

File: sdk/transaction.go [hunk 15]
```diff
@@ -9,22 +9,18 @@ import (
...
 
 type TransactionService struct {
 	*service
 	BlockchainService *BlockchainService
 }
 
-// returns Transaction for passed transaction id or hash
-func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (Transaction, error) {
+// returns Transaction for passed transaction group and hash
+func (txs *TransactionService) getTransaction(ctx context.Context, group string, id string) (Transaction, error) {
```
Comment (Wondertan) [inline sdk/transaction.go:20]: Change `group` to designated group type instead of a string.

File: sdk/transaction.go (hunk 11 above)
Comment (Wondertan) [reply sdk/transaction.go:91]: Agree with @xgreenx. Any user can simply use the generic Groups type himself to identify the needed transaction type.

@xgreenx pls help @Cempl to identify the exact methods that have to be removed and/or changed

File: sdk/transaction.go [hunk 16]
```diff
@@ -36,6 +32,106 @@ func (txs *TransactionService) GetTransaction(ctx context.Context, id string) (T
...
+
+// returns Transaction for passed transaction id or hash
+func (txs *TransactionService) GetAnyTransactionById(ctx context.Context, id string) (Transaction, error) {
+	trS, err := txs.GetTransactionStatus(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	return txs.getTransaction(ctx, trS.Group, id)
+}
```
Comment (Wondertan) [inline sdk/transaction.go:88]: @xgreenx, is it ok to do two requests to get any transactions??? Seems like a design flaw

File: sdk/transaction_internal.go [hunk 17]
```diff
@@ -41,3 +47,1119 @@ func (t *transactionStatusDTOs) toStruct() ([]*TransactionStatus, error) {
 
 	return statuses, nil
 }
+
```
Comment (Wondertan) [inline sdk/transaction_internal.go:50]: @xgreenx @Cempl, do I need to review this file? It seems like DTOs were simply moved without any other changes or am I missing something?

**Review by Wondertan — CHANGES_REQUESTED** [review]

**Comment by Wondertan** [comment]
@xgreenx, pls note that the PR contains API breaking changes, thus a minor(middle) version number has to be incremented, as per semver.

File: sdk/transaction.go (hunk 13 above)
Comment (Cempl) [reply sdk/transaction.go:136]: @Wondertan added [here](https://github.com/proximax-storage/go-xpx-chain-sdk/pull/85/commits/4072c37e8e379ac1458b31ccc30e97c313d9fde6) )

File: sdk/transaction.go (hunk 15 above)
Comment (Cempl) [reply sdk/transaction.go:20]: I will add a common method for it, because we have a lot of transaction groups and regarding task we need to add just a few groups, not all )

File: sdk/transaction_internal.go (hunk 17 above)
Comment (Cempl) [reply sdk/transaction_internal.go:50]: Please do not spend a time to review it, just moved files without changes ))

File: sdk/transaction.go [hunk 18]
```diff
@@ -72,7 +163,7 @@ func (txs *TransactionService) AnnounceAggregateBonded(ctx context.Context, tx *
 		tx.Payload,
 		tx.Hash.String(),
 	}
-	return txs.announceTransaction(ctx, &dto, announceAggregateRoute)
+	return txs.announceTransaction(ctx, &dto, fmt.Sprintf(transactionsByGroupRoute, partial.String()))
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply sdk/transaction.go:131]: Yeep

**Review by Wondertan — APPROVED** [review]
That's approval from my side considering all the missed points will be resolved

File: sdk/transaction.go (hunk 16 above)
Comment (Wondertan) [reply sdk/transaction.go]: @xgreenx, I think we need it in SDK, but the question is: Why do we need the status of a transaction while we already have its Hash/ID?

File: sdk/transaction.go [hunk 19]
```diff
@@ -85,11 +122,11 @@ func (txs *TransactionService) AnnounceAggregateBondedCosignature(ctx context.Co
 	return txs.announceTransaction(ctx, &dto, announceAggregateCosignatureRoute)
 }
 
-// returns TransactionStatus for passed transaction id or hash
+// GetTransactionStatus returns TransactionStatus for passed transaction id or hash
 func (txs *TransactionService) GetTransactionStatus(ctx context.Context, id string) (*TransactionStatus, error) {
```
Comment (xgreenx) [inline sdk/transaction.go:126]: Change the type of field "Group" in TransactionStatus from string to TransactionGroup please=) And use group here

**Review by xgreenx — APPROVED** [review]
